### PR TITLE
feat: scheduler (17/): add more scheduler logic

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR is part of the PRs that implement the Fleet workload scheduling.

It features more logic for scheduling PickN type CRPs.

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

-[x] Unit tests

### Special notes for your reviewer

To control the size of the PR, certain unit tests are not checked in; they will be sent in a separate PR.

### CODE REVIEW

File: pkg/scheduler/framework/frameworkutils.go [hunk 1]
```diff
@@ -346,3 +346,75 @@ func sortByClusterScoreAndName(bindings []*fleetv1beta1.ClusterResourceBinding)
...
+//     on the batch size; and the actual number of bindings created/updated is equal to the desired
+//     batch size
+//     -> in this case, no immediate requeue is necessary as all the work has been completed.
+//   - the desired batch size is equal to the batch size limit, i.e., no plugin has imposed a limit
+//     on the batch size; but the actual number of bindings created/updated is less than the desired
+//     batch size
+//     -> in this case, no immediate requeue is necessary as retries will not correct the situation;
+//     the scheduler should wait for the next signal from scheduling triggers, e.g., new cluster
+//     joined, or scheduling policy is updated.
+//   - the desired batch size is less than the batch size limit, i.e., a plugin has imposed a limit
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/frameworkutils.go:404]:
```suggestion
//   - the desired batch size is greater than the batch size limit, i.e., a plugin has imposed a limit
```

File: pkg/scheduler/framework/frameworkutils.go [hunk 2]
```diff
@@ -346,3 +346,75 @@ func sortByClusterScoreAndName(bindings []*fleetv1beta1.ClusterResourceBinding)
...
+//     on the batch size; and the actual number of bindings created/updated is equal to batch size
+//     limit
+//     -> in this case, immediate requeue is needed as there might be more fitting clusters to bind
+//     resources to.
+//   - the desired batch size is less than the batch size limit, i.e., a plugin has imposed a limit
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/frameworkutils.go:409]:
```suggestion
//   - the desired batch size is greater than the batch size limit, i.e., a plugin has imposed a limit
```

File: pkg/scheduler/framework/framework.go [hunk 3]
```diff
@@ -729,15 +729,87 @@ func (f *framework) runSchedulingCycleForPickNPlacementType(
...
+
+	// Extract the patched bindings.
+	patched := make([]*fleetv1beta1.ClusterResourceBinding, 0, len(toPatch))
+	for _, p := range toPatch {
+		patched = append(patched, p.updated)
+	}
+
+	// Update policy snapshot status with the latest scheduling decisions and condition.
+	klog.V(2).InfoS("Updating policy snapshot status", "clusterSchedulingPolicySnapshot", policyRef)
+	if err := f.updatePolicySnapshotStatusFrom(ctx, policy, filtered, toCreate, patched, scheduled, bound); err != nil {
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/framework.go:807]: should we check if we fully satisfy the placement requirement for the pickN?

Comment (michaelawyu) [reply pkg/scheduler/framework/framework.go:807]: This change will be sent in another PR; at this moment it still follows the old logic for simplicity reasons.

File: pkg/scheduler/framework/frameworkutils.go (hunk 1 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go]: Done.

File: pkg/scheduler/framework/frameworkutils.go (hunk 2 above)
Comment (michaelawyu) [reply pkg/scheduler/framework/frameworkutils.go]: Done.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**